lib/full_url_for.js

```diff
@@ -1,25 +1,18 @@
 'use strict';
 
-const { URL } = require('url');
+const { parse, URL } = require('url');
 const encodeURL = require('./encode_url');
-
-const urlObj = (str) => {
-  try {
-    return new URL(str);
-  } catch (err) {
-    return str;
-  }
-};
+const isExternalLink = require('./is_external_link');
 
 function fullUrlForHelper(path = '/') {
   if (path.startsWith('//')) return path;
+
   const { config } = this;
-  const data = urlObj(path);
+  const sitehost = parse(config.url).hostname || config.url;
+  const data = new URL(path, `http://${sitehost}`);
 
-  // Exit if this is an external path
-  if (typeof data === 'object') {
-    if (data.origin !== 'null') return path;
-  }
+  // Exit if this is an external link
+  if (isExternalLink.call(this, path)) return path;
```

Comment:
Emmmm. I don't think it is a good idea to call isExternalLink here.
url_for is used to complete the relative path while full_url_for not only complete relative path but also add domain prefix to make a absolute url.

Reply:
`isExternalLink()` takes `external_link.exclude` into consideration while `url_for` and `full_url_for` definitely don't need. So it becomes quite a clear. We can simply use `if (data.hostname !== sitehost) return path;` and it will be all fine.

Reply:
```suggestion
  if (data.origin !== 'null' || data.hostname !== sitehost) return path;
```

https://github.com/hexojs/hexo-util/blob/34088245fc4f576321546a7dab4cea10eb8ffd74/lib/is_external_link.js#L20
https://github.com/hexojs/hexo-util/blob/34088245fc4f576321546a7dab4cea10eb8ffd74/lib/is_external_link.js#L36

The main logic of `isExternalLink()` without `config.external_link.exclude`